### DISCUSSION TITLE
ISA NICs BootROM Window/Page/Bank Size

### DISCUSSION THREAD

**Discussion by chrisloweau** [ipxe/ipxe#1338] [General]
I can successfully compile and make iPXE 64KB ROMs for my old Intel and 3COM ISA NICs. However, the ROMs keep failing at boot - 3COM diags tell me the ROM has an invalid checksum, and using the old DEBUG command I discovered only the first 16KB of the ROMs appears in memory.

After much further research, it appears many ISA NICs only use a single 16KB UMA window, and then bank switch to access the total capacity of the ROM (up to 512KB and even 1MB apparently). A bit like how Expanded Memory works.

This means the firmware needs to be written to support a 16KB window and bank switching, which I can't find any indication iPXE is. The .mrom option sort of seem like the right direction, but doesn't appear to work for ISA NICs.

I've also gone back to try EtherBoot 5.4.4 and can compile smaller 32KB ROMs, but still only the first 16KB is mapped to UMA, and still fails.

And to confirm, the NICs are configured for the appropriate ROM sizes in their config utilities, and they are set for addressing that does not conflict with anything else. I have not found any setting for changing 16KB window size to the full ROM size (say 32 or 64KB). The particular NICs are Intel EtherExpress Pro 10 (eepro.isarom) and 3COM 3C515 (3c515.isarom).

I can't understand how iPXE (or EtherBoot) has ever supported ISA NICs!? Or have I got something in the NIC config completely wrong?

TIA
Chris

**Comment by jonathan-sturges** [comment]
Hi Chris,
Thanks for bringing this issue up.  I played with building 64k iPXE ROMs for an ISA RTL8019AS NIC a few years ago and didn't have any success either.  This week, I've also been testing PCI NICs (3c509b and tulip) on a 486 without success either.  I'm still interested because I like the idea of being able to inventory, triage and rescue older systems via netbooting.  It is frustrating, because NICs like the eepro have a 32-pin socket, implying at least a 1Mbit chip should work.  It doesn't seem unreasonable to just set the ROM size, enable & go, as you say.
If you ever find a particular ISA NIC that works without bank switching, please share, and I will do the same.  Thank you.

**Comment by mcb30** [comment]
It's an interesting issue.  By the time I started getting involved with Etherboot, network cards were pretty much all PCI.  The ISA ROM prefix works in emulation (where there is no bank switching), but I don't think I've ever tried putting iPXE in a real ISA NIC ROM.

As @chrisloweau says, the .mrom gives an indication of how this could be made to work.  iPXE already has support for building an image that looks like a very short stub ROM as far as the BIOS is concerned, with the `open_payload` subroutine called to allow the early-execution parts of iPXE to access the remainder of the iPXE image.

The API for `open_payload` is that it gets passed a buffer address in `%esi` that is large enough to hold the complete iPXE image, with `%ecx` being the offset within that buffer of the first "non-early" block (i.e. the first block that isn't part of the stub ROM image), and it has to return with a copy of the complete iPXE image contiguous in memory and pointed to by `%esi + %ecx`.

There are three current implementations of `open_payload`:
* The default implementation provided by a weak symbol in `romprefix.S`, which just returns immediately (i.e. assumes that there are no clever schemes involved, and that the ROM image as seen by the BIOS already contains all of the blocks of the complete iPXE image).
* The trivial implementation in `lkrnprefix.S`, which knows that the remainder of the image will have been loaded at the fixed bzImage address of 0x100000 and so just returns that fixed address.
* The implementation in `mromprefix.S`, which does a lot of PCI BAR wizardry to temporarily expose the expansion ROM BAR and then copies the contents to the buffer at `%esi`.

So, for ISA ROMs with paging mechanisms, you'd need a separate `promprefix.S` (or whatever) that similarly sets `ROMPREFIX_EXCLUDE_PAYLOAD` to create a stub ROM, and provides an `open_payload` implementation that uses the ISA ROM's paging mechanism to access each page in turn and copy it to the provided buffer at `%esi`.

There is then the separate issue of how this buffer is actually allocated in `romprefix.S`.  The API for `open_payload` currently makes this an **optional** buffer: the implementation of `open_payload` is allowed to copy content to the buffer as a performance optimisation if the buffer exists, but it has to return a valid pointer in `%esi + %ecx` even if no buffer was provided.  This makes sense for the current use cases:
* `mromprefix.S` has the option of just pointing `%esi` to the mapped expansion ROM BAR, which will be slow but still functional.
* `lkrnprefix.S` always ignores the buffer and points `%esi` to the fixed load address at 0x100000 anyway
The paging ISA ROM prefix wouldn't have the ability to work without a buffer address, since there isn't **any** valid physical address that can be used to access the contiguous iPXE image without copying.  So, the semantics of that would have to change.  For testing during development, if your system doesn't support PMM and so can't allocate a suitable block, it would be adequate to just hardcode a `mov $0x100000, %esi` to always put the copy at the 1MB.

There is already logic in `libprefix.S` for allocating a temporary block of high memory for the decompression area, which could be fairly easily extended to provide a temporary block large enough to hold both the image source copy and the decompression area, but that's something to address after getting the basic paging ISA ROM prefix working.

Good luck!
